### PR TITLE
Add sharp bits section and invite users to the discussions upon error

### LINKED ISSUE #60: An overview of "sharp bits" of RxInfer
This is inspired by [JAX Sharp Bits](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html) 

It would be awesome to have a documentation page describing the most unexpected differences to what a normal Julia user (or a Turing.jl user) would expect, eg,
- what the allowed operations are within a model (you cannot use functions // you cannot use sum on r a random variable array // you cannot use `+` with more than 2 arguments)
- what are the initialization needs and the minimum required
- troubleshooting frequently occurring problems
    - what should be in your function return statement (because you cannot just execute the model anyway. my guess: it's for the quantities that you want to manually subscribe to)
    - if your fit is bad at the very beginning of the time series, try increasing iterations
    - if you have BFE NaNs check that there isn't any input multiplied by a MvNormal with only zeros; if there is add, add `tiny` to the first column
    - if you get MethodError with q_ m_, try XYZ first

### PR DESCRIPTION
**Pull request by bvdmitri**
Closes #60 #385 

This was on our issue tracker for a while, but with the help of Cursor things get easier. Please review and suggest changes.

This pull request introduces several updates to the RxInfer documentation, focusing on improving user guidance and addressing common issues. The changes include adding a new "Sharp bits of RxInfer" section, updating the documentation to include references to this new section, and enhancing error handling and messaging within the inference process.

Documentation improvements:

* [`docs/make.jl`](diffhunk://#diff-4aae2d1c783cade58bd2cb13748da956e568b7f2aed5fafd9e2a46fb97daf613R88-R91): Introduced a draft mode for documentation generation, and added a new section "Sharp bits of RxInfer" to the documentation index. [[1]](diffhunk://#diff-4aae2d1c783cade58bd2cb13748da956e568b7f2aed5fafd9e2a46fb97daf613R88-R91) [[2]](diffhunk://#diff-4aae2d1c783cade58bd2cb13748da956e568b7f2aed5fafd9e2a46fb97daf613L114-R122)
* [`docs/src/manuals/comparison.md`](diffhunk://#diff-1fa7cc8af5973e431f7c395387c61cba1d2ea023df304a20cc37c778cd874602L70-R75): Added a note about using `:=` for deterministic relationships in RxInfer.
* [`docs/src/manuals/model-specification.md`](diffhunk://#diff-a8e298280fb777297db4c150dd97524a805bb77fd987a919e4b83c6f9bd31302L174-R176): Included a reference to the new "Sharp bits" section for further explanation on deterministic nodes.

New "Sharp bits of RxInfer" section:

* [`docs/src/manuals/sharpbits/overview.md`](diffhunk://#diff-6b8bbf209f29ae77fe103c4cf707c1d1382df5b62f0501a97341ef148e6193adR1-R50): Created an overview page for common pitfalls and issues in RxInfer, along with guidelines for contributing to this section.
* [`docs/src/manuals/sharpbits/rule-not-found.md`](diffhunk://#diff-cdd8ec5405c44e0786b2860dd162e238d97f2920d8a628d99cac8dbb3fedc2a0R1-R100): Added detailed documentation on handling `RuleNotFoundError`, including causes, common scenarios, and solutions.
* [`docs/src/manuals/sharpbits/stack-overflow-inference.md`](diffhunk://#diff-d75f4167290d36952e100eed870daee63cccb815ddc47b644e555ec822ed9dd2R1-R97): Provided guidance on preventing `StackOverflowError` during inference, with examples and performance considerations.
* [`docs/src/manuals/sharpbits/usage-colon-equality.md`](diffhunk://#diff-b5b5c49d96b0a6a1061ef2fbabda2e2ba0b90cd40fe5140bb5957ed764f4aa3dR1-R83): Explained the use of `:=` for deterministic nodes and the reasoning behind not using `=`.

Enhanced error handling:

* [`src/inference/inference.jl`](diffhunk://#diff-fb9a53dd91325aaa35b311d6efc7eccacd1c187c08e9ffe5e0d813152f076308L95-R135): Improved error messaging for `StackOverflowError` during inference, including suggestions for resolving the issue and links to relevant documentation. [[1]](diffhunk://#diff-fb9a53dd91325aaa35b311d6efc7eccacd1c187c08e9ffe5e0d813152f076308L95-R135) [[2]](diffhunk://#diff-fb9a53dd91325aaa35b311d6efc7eccacd1c187c08e9ffe5e0d813152f076308R251-R253)

### CODE REVIEW

**Comment by abpolym** [comment]
I'll check it out later today. I assume I can create the docs locally with `make docs`?

**Comment by bvdmitri** [comment]
yes `make docs` command should be enough

**Comment by abpolym** [comment]
`make docs` first threw warnings that examples at `docs/src/examples` do not exist, so I had to make them with `make examples`.

Before that, it threw an error:
`Error: failed to run `@example` block in src/manuals/sharpbits/stack-overflow-inference.md:31-5`.

`make docs` works now, I'll check the documentation afterwards.

**Comment by bvdmitri** [comment]
> make docs first threw warnings that examples at docs/src/examples do not exist, so I had to make them with make examples.
yeah, its a warning, not an error, it will work without examples just fine, I also don't have them since they take a lot of time to compute

> It continued by throwing an error: Error: failed to run @example block in src/manuals/sharpbits/stack-overflow-inference.md:31-5.

Well this is interesting, I don't have this error, perhaps there is a bug or smth

> I have no idea what I have to do to fix those errors for either Julia version

Well this is jammer, I don't have these errors and CI seems to work just fine too, not sure what can be the reason... :(

File: src/inference/inference.jl [hunk 1]
```diff
@@ -92,24 +92,47 @@ function inference_process_error(error)
...
+    Here are some helpful resources to get you back on track:
+
+    1. Check our Sharp bits documentation which covers common issues:
+       https://reactivebayes.github.io/RxInfer.jl/stable/manuals/sharpbits/overview/
+
+    2. Browse our existing issues - your question may already be answered:
+       https://github.com/ReactiveBayes/RxInfer.jl/issues
+
+    Still stuck? We'd love to help! You can:
+    - Start a discussion for questions and help:
```
Comment (fonsp) [inline src/inference/inference.jl:121]: Super nice!! ✨

```suggestion
    - Start a discussion for questions and help. Feedback and questions from new users is also welcome! If you are stuck, please reach out and we will solve it together.
```

Maybe some more words to fight the idea that GH discussions is only for:
- experts
- bugs/technical problems

**Comment by bvdmitri** [comment]
> Error: renderer for julia:cairo is unavailable 

@abpolym Indeed https://discourse.julialang.org/t/issue-occurring-randomly-with-1-11/121233/4
also happens on CI now https://github.com/ReactiveBayes/RxInfer.jl/actions/runs/12463475091/job/34786844959

File: docs/src/manuals/sharpbits/rule-not-found.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+2. Rules for combining these messages based on the factor `f`'s type
+3. Rules for producing the outgoing message type that `y` expects
+
+A `RuleNotFoundError` occurs when any of these rules are missing. For example, if `x` sends a `Normal` message but `f` doesn't know how to process `Normal` inputs, or if `f` can't produce the type of message that `y` expects.
+
+## Solutions
+
+### 1. Convert to conjugate pairs
+
+First, try to reformulate your model using conjugate prior-likelihood pairs. Conjugate pairs have analytical solutions for message passing and are well-supported in RxInfer. For example, instead of using a Normal likelihood with Beta prior on its precision, use a Normal-Gamma conjugate pair. See [Conjugate prior - Wikipedia](https://en.wikipedia.org/wiki/Conjugate_prior#Table_of_conjugate_distributions) for a comprehensive list of conjugate distributions.
```
Comment (abpolym) [inline docs/src/manuals/sharpbits/rule-not-found.md:91]: You've been using code blocks for distributions like `Normal`, `Beta` everywhere but here 😄 .

File: docs/src/manuals/sharpbits/stack-overflow-inference.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+## Solution: Limiting Stack Depth
+
+RxInfer provides a solution through the `limit_stack_depth` option in the inference options. This option limits the recursion depth at the cost of some performance overhead.
+
+### How to Use
+
+You can enable stack depth limiting by passing it through the `options` parameter to the `infer` function:
+
+```@example stack-overflow-inference
```
Comment (abpolym) [inline docs/src/manuals/sharpbits/stack-overflow-inference.md:31]: Is it possible to use julia code blocks?

I think if I just add "julia" right after the three dashes "`" but before the (at)example then it should use color highlighting?

Test:
```julia @example stack-overflow-inference
using RxInfer
@model function long_state_space_model(y)
    x[1] ~ Normal(mean = 0.0, var = 1.0)
    y[1] ~ Normal(mean = x[1], var = 1.0)
    for i in 2:length(y)
        x[i] ~ Normal(mean = x[i - 1], var = 1.0)
        y[i] ~ Normal(mean = x[i], var = 1.0)
    end
end
data = (y = rand(10000), )
using Test #hide
@test_throws StackOverflowError infer(model = long_state_space_model(), data = data) #hide
results = infer(
    model = long_state_space_model(),
    data = data,
    options = (
        limit_stack_depth = 100, # note the comma
    )
)
```

File: docs/src/manuals/sharpbits/stack-overflow-inference.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+- Dealing with long sequences or time series
+
+!!! tip
+    If you're not sure whether you need this option, try running your model without it first. Only enable `limit_stack_depth` if you encounter stack overflow issues.
+
+## Further Reading
+
+For more details about inference options and execution, see:
+- [Static Inference](@ref manual-static-inference) documentation
+- The `options` parameter in the [`infer`](@ref) function documentation
```
Comment (abpolym) [inline docs/src/manuals/sharpbits/stack-overflow-inference.md:95]: I assume these references will render properly in HTML, in markdown it is broken 😢

**Review by abpolym — COMMENTED** [review]
Still need to be able to see the proper rendering, but the changes look very nice/informative so far.

File: docs/src/manuals/sharpbits/stack-overflow-inference.md (hunk 3 above)
Comment (bvdmitri) [reply docs/src/manuals/sharpbits/stack-overflow-inference.md:31]: it uses highlighted julia code blocks with `@example stack-overflow-inference`, see [here](https://documenter.juliadocs.org/stable/man/syntax/#reference-at-example), these are julia code blocks that are also being executed during the creation of the documentation page, just `julia` will only highlight but not execute

File: docs/src/manuals/sharpbits/stack-overflow-inference.md (hunk 4 above)
Comment (bvdmitri) [reply docs/src/manuals/sharpbits/stack-overflow-inference.md:95]: yes because there are not from markdown, but from Documenter.jl

**Comment by bvdmitri** [comment]
Thanks @abpolym!

**Review by albertpod — APPROVED** [review]
Given the incorporation of suggested changes from others, LGTM

**Pull request merged**